src/value_generators/src/DocumentGenerator.cpp

```diff
@@ -1113,6 +1116,108 @@ class FastRandomStringGenerator : public StringGenerator {
     }
 };
 
+// The following is a replacement for partial template specialization for functions, which is not
+// currently supported in C++.
+template <typename F, typename T>
+struct convert{};
+
+// Conversion used between int, long, double
+template <typename F, typename T>
+T convertToNumeric(convert<F, T>, const F& from)
+{
+    return (T)from;
+}
+
+// Special conversion for decimal
+template<typename F>
+bsoncxx::decimal128 convertToNumeric(convert<F, bsoncxx::decimal128>, const F& from) {
+    // mongocxx doesn't define nice converters from numeric types to decimal, so we convert first
+    // to string and then to decimal
+    return bsoncxx::decimal128(std::to_string(from));
+}
+
+// String conversions
+bsoncxx::decimal128 convertToNumeric(convert<std::string, bsoncxx::decimal128>,
+        const std::string& s) {
+    return bsoncxx::decimal128(s);
+}
+double convertToNumeric(convert<std::string, double>, const std::string& s) {
+    return stod(s);
+}
+int64_t convertToNumeric(convert<std::string, int64_t>, const std::string& s) {
+    return stol(s);
+}
+int32_t convertToNumeric(convert<std::string, int32_t>, const std::string& s) {
+    return stoi(s);
+}
+
+// This is the only overload we run directly, and it will route to the correct template.
+template <typename F, typename T>
+T convertToNumeric(const F& from) {
+    return convertToNumeric(convert<F, T>{}, from);
+}
+
+// overloaded struct for use with std::variants.
+template<class... Ts>
+struct overloaded : Ts... { using Ts::operator()...; };
+
+template<class... Ts>
+overloaded(Ts...) -> overloaded<Ts...>;
+
+// Generator to convert from any numeric type except decimal or string to specific numeric type.
+template<typename T>
+class NumericConversionGenerator : public Generator<T> {
+public:
+    NumericConversionGenerator(const Node& node, GeneratorArgs generatorArgs) {
+        // Try all of the possible generators for from, return when we find one that works.
+        try {
+            _generator = stringGenerator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        try {
+            _generator = intGenerator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        try {
+            _generator = doubleGenerator(node["from"], generatorArgs);
+            return;
+        } catch (const InvalidConversionException& e) {}
+          catch (const UnknownParserException& e) {}
+        try {
+            _generator = int32Generator(node["from"], generatorArgs);
```

Comment:
I don't think this catches ambiguous values that could parse as multiple options. Do we want to try them all and then barf if more than one works?

Reply:
Done; however I added a special case for constant values, as those can parse successfully as multiple types (for example, `{^ConvertToInt: {from: "1"}}` can parse the "from" node successfully with any generator, because of how `Node::to<T>` operates), and can always be parsed as strings.